x-pack/libbeat/autodiscover/providers/aws/elb/lblistener.go

```diff
@@ -0,0 +1,66 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package elb
+
+import (
+	"github.com/aws/aws-sdk-go-v2/service/elasticloadbalancingv2"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+// lbListener is a tuple type representing an elasticloadbalancingv2.Listener and its associated elasticloadbalancingv2.LoadBalancer.
+type lbListener struct {
+	lb       *elasticloadbalancingv2.LoadBalancer
+	listener *elasticloadbalancingv2.Listener
+}
+
+// toMap converts this lbListener into the form consumed as metadata in the autodiscovery process.
+func (l *lbListener) toMap() common.MapStr {
+	m := common.MapStr{}
+
+	// We fully spell out listener_arn to avoid confusion with the ARN for the whole ELB
+	m["listener_arn"] = l.listener.ListenerArn
+	m["load_balancer_arn"] = *l.lb.LoadBalancerArn
+	m["host"] = *l.lb.DNSName
+	m["port"] = *l.listener.Port
+	m["protocol"] = l.listener.Protocol
+	m["type"] = string(l.lb.Type)
+	m["scheme"] = l.lb.Scheme
+	m["availability_zones"] = l.azStrings()
+	m["created"] = l.lb.CreatedTime
+	m["state"] = l.stateMap()
+	m["ip_address_type"] = string(l.lb.IpAddressType)
+	m["security_groups"] = l.lb.SecurityGroups
+	m["vpc_id"] = *l.lb.VpcId
```

Comment:
We should check if any of these derefereced variables are nil, if they are for any reason this will panic. (Is it possible to have a load balancer without DNS name for example?)

Reply:
I've made this safer. There is always a DNS name set by amazon. The AWS lib just always uses `*string` instead of `string` for reasons that are unclear to me 🤷‍♂ . I assume they think it's more efficient memory-wise.